eZ/Publish/Core/Persistence/Cache/InMemory/InMemoryCache.php

```diff
@@ -0,0 +1,203 @@
+ *
+ * Allow to use one instance with low ttl (200-600 milliseconds) and limited items (~100) dedicated to content (and
+ * related classes) to handle burst caching, and a separate instance with longer ttl (3-10sec) and separate limit (~200)
+ * for meta values like content type, section, ... to handle repeated lookups for these seldom changing values.
+ *
+ * @internal Only for use in eZ\Publish\Core\Persistence\Cache\AbstractInMemoryHandler, may change depending on needs there.
+ */
+class InMemoryCache
+{
+    /**
+     * @var float Cache Time to Live, in seconds. This is only for how long we keep cache object around in-memory.
+     */
+    private $ttl;
+
+    /**
+     * @var int The limit of objects in cache pool at a given time
+     */
+    private $limit;
+
+    /**
+     * @var bool Switch for enabeling/disabling in-memory cache
+     */
+    private $enabled;
+
+    /**
+     * Cache objects by primary key.
+     *
+     * @var object[]
+     */
+    private $cache = [];
+
+    /**
+     * @var float[] Timestamp (float microtime) for individual cache by primary key.
+     */
+    private $cacheTime = [];
+
+    /**
+     * Mapping of secondary index to primary key.
+     *
+     * @var string[]
+     */
+    private $cacheIndex = [];
+
+    /**
+     * @var float|null Micro timestamp with time clear($global=true) has been called to synchronize across cache pools.
+     */
+    private $cacheExpiryTime;
+    protected static $globalCacheExpiry;
+
+    /**
+     * Language Cache constructor.
+     *
+     * @param float $ttl Seconds for the cache to live as a float, by default 0.3 (300 milliseconds)
+     * @param int $limit Limit for values to keep in cache, by default 100 cache values (per pool instance).
+     * @param bool $enabled For use by configuration to be able to disable or enable depending on needs.
+     */
+    public function __construct(float $ttl = 0.3, int $limit = 100, bool $enabled = true)
+    {
+        $this->ttl = $ttl;
+        $this->limit = $limit;
+        $this->enabled = $enabled;
+    }
+
+    /**
+     * Returns a cache objects.
+     *
+     * @param string $key Primary or secondary index to look for cache on.
+     *
+     * @return object|null Object if found, null if not.
+     */
+    public function get(string $key)
+    {
+        if ($this->enabled === false) {
+            return null;
+        }
+
+        // Check for global expiry change, if the case clear cache
+        if ($this->cacheExpiryTime !== self::$globalCacheExpiry) {
+            $this->clear();
+            $this->cacheExpiryTime = self::$globalCacheExpiry;
+
+            return null;
+        }
+
+        $index = $this->cacheIndex[$key] ?? $key;
+        if (!isset($this->cache[$index]) || $this->cacheTime[$index] + $this->ttl < microtime(true)) {
+            return null;
+        }
+
+        return $this->cache[$index];
```

Comment:
maybe we could add a call to an empty method here, to trace request time / number of accesses. This way subclasses could implement RLU and friends for vacuum?

Reply:
Then imho we should rather add a interface at some point so people can add own impl, psr-16 is such an interface but for now I don't bother as it has [a lot more methods then we need atm](https://github.com/php-fig/simple-cache/pull/15).